<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор" data-toc-modified-id="Обзор-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a href="#Посмотрим-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посмотрим-сколько-книг-вышло-после-1-января-2000-года-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Посмотрим сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-для-каждой-книги-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Количество обзоров и средняя оценка для каждой книги</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-исключим-из-анализа-брошюры" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-исключим-из-анализа-брошюры-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так исключим из анализа брошюры</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li><li><span><a href="#Итог" data-toc-modified-id="Итог-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Итог</a></span></li></ul></div>

# Знакомство с базой данных книжного сервиса

Компания купила крупный сервис для чтения книг по подписке. Теперь первой задачей станет - проанализировать базу данных.  

**Задачи:**  
* Посчитайте, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитайте количество обзоров и среднюю оценку;
* Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
* Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
* Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Описание данных**  

Таблица **books**
Содержит данные о книгах:  
`book_id` — идентификатор книги;  
`author_id` — идентификатор автора;  
`title` — название книги;  
`num_pages` — количество страниц;  
`publication_date` — дата публикации книги;  
`publisher_id` — идентификатор издателя.  

Таблица **authors**  
Содержит данные об авторах:  
`author_id` — идентификатор автора;  
`author` — имя автора.  

Таблица **publishers**  
Содержит данные об издательствах:  
`publisher_id` — идентификатор издательства;  
`publisher` — название издательства;  

Таблица **ratings**  
Содержит данные о пользовательских оценках книг:  
`rating_id` — идентификатор оценки;  
`book_id` — идентификатор книги;  
`username` — имя пользователя, оставившего оценку;  
`rating` — оценка книги.  

Таблица **reviews**  
Содержит данные о пользовательских обзорах на книги:  
`review_id` — идентификатор обзора;  
`book_id` — идентификатор книги;  
`username` — имя пользователя, написавшего обзор;  
`text` — текст обзора.

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine 

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Обзор

Исследуем таблицы — выведем первые строки, посчитаем количество строк
в каждой таблице;

### Таблица books

In [3]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT COUNT(*) OVER () AS count_rows,
                   *
           FROM books 
           LIMIT 5;'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count_rows  book_id  author_id  \
0        1000        1        546   
1        1000        2        465   
2        1000        3        407   
3        1000        4         82   
4        1000        5        125   

                                               title  num_pages  \
0                                       'Salem's Lot        594   
1                 1 000 Places to See Before You Die        992   
2  13 Little Blue Envelopes (Little Blue Envelope...        322   
3  1491: New Revelations of the Americas Before C...        541   
4                                               1776        386   

  publication_date  publisher_id  
0       2005-11-01            93  
1       2003-05-22           336  
2       2010-12-21           135  
3       2006-10-10           309  
4       2006-07-04           268

Таблица содержит данные о 1000 книгах.

### Таблица authors

In [4]:
query = '''SELECT COUNT(*) OVER () AS count_rows,
                   *
           FROM authors 
           LIMIT 5;'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count_rows  author_id                          author
0         636          1                      A.S. Byatt
1         636          2  Aesop/Laura Harris/Laura Gibbs
2         636          3                 Agatha Christie
3         636          4                   Alan Brennert
4         636          5        Alan Moore/David   Lloyd

Таблица содержит данные о 636 авторов.

### Таблица publishers

In [5]:
query = '''SELECT COUNT(*) OVER () AS count_rows,
                   *
           FROM publishers 
           LIMIT 5;'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count_rows  publisher_id                          publisher
0         340             1                                Ace
1         340             2                           Ace Book
2         340             3                          Ace Books
3         340             4                      Ace Hardcover
4         340             5  Addison Wesley Publishing Company

Таблица содержит данные о 340 издательствах.

### Таблица ratings

In [6]:
query = '''SELECT COUNT(*) OVER () AS count_rows,
                   *
           FROM ratings 
           LIMIT 5;'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count_rows  rating_id  book_id       username  rating
0        6456          1        1     ryanfranco       4
1        6456          2        1  grantpatricia       2
2        6456          3        1   brandtandrea       5
3        6456          4        2       lorichen       3
4        6456          5        2    mariokeller       2

Таблица содержит 6456 записей о рейтингах книг.

### Таблица reviews

In [7]:
query = '''SELECT COUNT(*) OVER () AS count_rows,
                   *
           FROM reviews 
           LIMIT 5;'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count_rows  review_id  book_id       username  \
0        2793          1        1   brandtandrea   
1        2793          2        1     ryanfranco   
2        2793          3        2       lorichen   
3        2793          4        3  johnsonamanda   
4        2793          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Таблица содержит 2793 записи о пользовательских обзорах на книги.

## Посмотрим сколько книг вышло после 1 января 2000 года

In [8]:
query = '''SELECT COUNT(book_id)
           FROM books 
           WHERE  publication_date > '2000-01-01';
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2000 года вышло 819 книг

##  Количество обзоров и средняя оценка для каждой книги

In [9]:
query = '''SELECT b.book_id,
                  b.title,
                  COUNT(text), 
                  AVG(rating)
           FROM books AS b
           LEFT JOIN ratings AS r ON r.book_id = b.book_id
           LEFT JOIN reviews AS rev ON rev.book_id = b.book_id
           GROUP BY b.book_id
           ORDER BY count DESC ;
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  count  \
0        948                            Twilight (Twilight  #1)   1120   
1        750                The Hobbit  or There and Back Again    528   
2        673                             The Catcher in the Rye    516   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...    492   
4        299  Harry Potter and the Chamber of Secrets (Harry...    480   
..       ...                                                ...    ...   
995      221                          Essential Tales and Poems      0   
996      808                            The Natural Way to Draw      0   
997      191  Disney's Beauty and the Beast (A Little Golden...      0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel      0   
999      672   The Cat in the Hat and Other Dr. Seuss Favorites      0   

          avg  
0    3.662500  
1    4.125000  
2    3.825581  
3    4.414634  
4    4.287500  
..        ...  
995  4.000000  
996  3.000000  
997  4.000000  
998  3.666667  
999  5.000000  

[1000 rows x 4 columns]

Некоторые книги содержат огромное количество обзоров, а некоторые ни одного. Но количество обзоров никак не влияет на средний рейтинг.

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так исключим из анализа брошюры

In [10]:
query = '''WITH top AS (SELECT publisher_id,
                               COUNT(book_id) AS book_cnt
                        FROM books
                        WHERE num_pages > 50
                        GROUP BY publisher_id
                        ORDER BY book_cnt DESC
                        LIMIT 1)
            SELECT p.publisher,
                   top.book_cnt
            FROM publishers AS p
            JOIN top ON p.publisher_id = top.publisher_id
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  book_cnt
0  Penguin Books        42

Больше всего книг выпустило издательство *Penguin Books* - 42 книги.

## Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
query = '''WITH
               name_1 AS (SELECT book_id, 
                            AVG(rating) AS book_rating
                     FROM ratings
                     GROUP BY book_id
                     HAVING COUNT(rating_id) >= 50
                    ), 
               name_2 AS (SELECT b.book_id, 
                              au.author, 
                              a.book_rating
                      FROM books AS b
                      JOIN name_1 AS a ON b.book_id = a.book_id
                      JOIN authors AS au ON au.author_id = b.author_id
                    )

               SELECT author, 
                      AVG(book_rating) AS avg_rating
               FROM name_2
               GROUP BY author
               ORDER BY avg_rating DESC
               LIMIT 1;
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

Среди популярных книг, самый высокий рейтинг книг у Д. Роулинг - 4,28  
В предыдущих заданиях мы видели, что есть книги и сболее высоким рейтингом. Но в данном обзоре мы учитываем только книги,у которых больше 50 оценок.

## Среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [12]:
query = '''WITH
               a AS (SELECT username
                     FROM ratings 
                     GROUP BY username
                     HAVING COUNT(rating_id) > 48
                    ),    
               b AS (SELECT COUNT(review_id) AS review_cnt
                     FROM reviews AS r
                     JOIN a ON r.username = a.username
                     GROUP BY a.username)
    
           SELECT AVG(review_cnt)
           FROM b
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 обзора.

## Итог

В базе данных содержатся 5 таблиц:  

`books` - имеет 6 столбцов и 1000 записей о книгах;  
`authors` - имеет 2 столбца и 656 записей об авторах;  
`publishers` - имеет 2 столбцa и 340 записей об издательствах;  
`ratings` - имеет 4 столбцa и  6456 записей о рейтингах книг;  
`reviews` - имеет 4 столбца и 2793 записи о пользовательских обзорах на книги.   

* После 1 января 2000 года вышло 819 книг
* Некоторые книги содержат огромное количество обзоров, а некоторые ни одного. Но количество обзоров никак не влияет на средний рейтинг. 
* Больше всего книг выпустило издательство Penguin Books - 42 книги.
* Самый большой рейтинг у автора Д. Роулинг (среди книги,у которых больше 50 оценок) - 4.3
* Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 обзора.